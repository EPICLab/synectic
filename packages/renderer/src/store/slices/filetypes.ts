import {createEntityAdapter, createSlice} from '@reduxjs/toolkit';
import type {Filetype} from '@syn-types/filetype';

export const filetypeAdapter = createEntityAdapter<Filetype>();

export const filetypeSlice = createSlice({
  name: 'filetypes',
  initialState: filetypeAdapter.getInitialState(),
  reducers: {
    filetypeAdded: filetypeAdapter.addOne,
    filetypeRemoved: filetypeAdapter.removeOne,
    filetypeUpdated: filetypeAdapter.upsertOne,
  },
});

export const {filetypeAdded, filetypeRemoved, filetypeUpdated} = filetypeSlice.actions;

export default filetypeSlice.reducer;
