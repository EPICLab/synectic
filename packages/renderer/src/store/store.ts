import {combineReducers, configureStore} from '@reduxjs/toolkit';
import branchesReducer from './slices/branches';
import cardsReducer from './slices/cards';
import commitsReducer from './slices/commits';
import filetypesReducer from './slices/filetypes';
import metafilesReducer from './slices/metafiles';
import modalsReducer from './slices/modals';
import reposReducer from './slices/repos';
import stacksReducer from './slices/stacks';
import {listenerMiddleware} from './listenerMiddleware';

export const rootReducer = combineReducers({
  branches: branchesReducer,
  cards: cardsReducer,
  commits: commitsReducer,
  filetypes: filetypesReducer,
  metafiles: metafilesReducer,
  modals: modalsReducer,
  repos: reposReducer,
  stacks: stacksReducer,
});

const store = configureStore({
  reducer: rootReducer,
  devTools: true,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware().concat([listenerMiddleware.middleware]),
});

export type RootState = ReturnType<typeof rootReducer>;
export type AppDispatch = typeof store.dispatch;

export default {store: store};
