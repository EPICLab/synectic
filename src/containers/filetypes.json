[
  {
    "filetype": "ABAP",
    "handler": "Editor",
    "extensions": [
      "abap"
    ]
  },
  {
    "filetype": "ActionScript",
    "handler": "Editor",
    "extensions": [
      "as"
    ]
  },
  {
    "filetype": "ADA",
    "handler": "Editor",
    "extensions": [
      "ada",
      "adb"
    ]
  },
  {
    "filetype": "Apache_Conf",
    "handler": "Editor",
    "extensions": [
      ".htaccess",
      ".htgroups",
      ".htpasswd",
      ".conf",
      "htaccess",
      "htgroups",
      "htpasswd"
    ]
  },
  {
    "filetype": "AsciiDoc",
    "handler": "Editor",
    "extensions": [
      "asciidoc",
      "adoc"
    ]
  },
  {
    "filetype": "Assembly_x86",
    "handler": "Editor",
    "extensions": [
      "asm",
      "a"
    ]
  },
  {
    "filetype": "AutoHotKey",
    "handler": "Editor",
    "extensions": [
      "ahk"
    ]
  },
  {
    "filetype": "BatchFile",
    "handler": "Editor",
    "extensions": [
      "bat",
      "cmd"
    ]
  },
  {
    "filetype": "Bro",
    "handler": "Editor",
    "extensions": [
      "bro"
    ]
  },
  {
    "filetype": "C_Cpp",
    "handler": "Editor",
    "extensions": [
      "cpp",
      "c",
      "cc",
      "cxx",
      "h",
      "hh",
      "hpp",
      "ino"
    ]
  },
  {
    "filetype": "C9Search",
    "handler": "Editor",
    "extensions": [
      "c9search_results"
    ]
  },
  {
    "filetype": "Cirru",
    "handler": "Editor",
    "extensions": [
      "cirru",
      "cr"
    ]
  },
  {
    "filetype": "Clojure",
    "handler": "Editor",
    "extensions": [
      "clj",
      "cljs"
    ]
  },
  {
    "filetype": "Cobol",
    "handler": "Editor",
    "extensions": [
      "CBL",
      "COB"
    ]
  },
  {
    "filetype": "coffee",
    "handler": "Editor",
    "extensions": [
      "coffee",
      "cf",
      "cson",
      "Cakefile"
    ]
  },
  {
    "filetype": "ColdFusion",
    "handler": "Editor",
    "extensions": [
      "cfm"
    ]
  },
  {
    "filetype": "CSharp",
    "handler": "Editor",
    "extensions": [
      "cs"
    ]
  },
  {
    "filetype": "Csound_Document",
    "handler": "Editor",
    "extensions": [
      "csd"
    ]
  },
  {
    "filetype": "Csound_Orchestra",
    "handler": "Editor",
    "extensions": [
      "orc"
    ]
  },
  {
    "filetype": "Csound_Score",
    "handler": "Editor",
    "extensions": [
      "sco"
    ]
  },
  {
    "filetype": "CSS",
    "handler": "Editor",
    "extensions": [
      "css"
    ]
  },
  {
    "filetype": "Curly",
    "handler": "Editor",
    "extensions": [
      "curly"
    ]
  },
  {
    "filetype": "D",
    "handler": "Editor",
    "extensions": [
      "d",
      "di"
    ]
  },
  {
    "filetype": "Dart",
    "handler": "Editor",
    "extensions": [
      "dart"
    ]
  },
  {
    "filetype": "Diff",
    "handler": "Editor",
    "extensions": [
      "diff",
      "patch"
    ]
  },
  {
    "filetype": "Directory",
    "handler": "Explorer",
    "extensions": []
  },
  {
    "filetype": "Dockerfile",
    "handler": "Editor",
    "extensions": [
      ".Dockerfile"
    ]
  },
  {
    "filetype": "Dot",
    "handler": "Editor",
    "extensions": [
      "dot"
    ]
  },
  {
    "filetype": "Drools",
    "handler": "Editor",
    "extensions": [
      "drl"
    ]
  },
  {
    "filetype": "DummySyntax",
    "handler": "Editor",
    "extensions": [
      "dummy"
    ]
  },
  {
    "filetype": "HTML",
    "handler": "Editor",
    "extensions": [
      "html",
      "htm",
      "xhtml",
      "vue",
      "we",
      "wpy"
    ]
  },
  {
    "filetype": "Eiffel",
    "handler": "Editor",
    "extensions": [
      "e",
      "ge"
    ]
  },
  {
    "filetype": "EJS",
    "handler": "Editor",
    "extensions": [
      "ejs"
    ]
  },
  {
    "filetype": "Elixir",
    "handler": "Editor",
    "extensions": [
      "ex",
      "exs"
    ]
  },
  {
    "filetype": "Elm",
    "handler": "Editor",
    "extensions": [
      "elm"
    ]
  },
  {
    "filetype": "Erlang",
    "handler": "Editor",
    "extensions": [
      "erl",
      "hrl"
    ]
  },
  {
    "filetype": "Forth",
    "handler": "Editor",
    "extensions": [
      "frt",
      "fs",
      "ldr",
      "fth",
      "4th"
    ]
  },
  {
    "filetype": "Fortran",
    "handler": "Editor",
    "extensions": [
      "f",
      "f90"
    ]
  },
  {
    "filetype": "FTL",
    "handler": "Editor",
    "extensions": [
      "ftl"
    ]
  },
  {
    "filetype": "Gcode",
    "handler": "Editor",
    "extensions": [
      "gcode"
    ]
  },
  {
    "filetype": "Gherkin",
    "handler": "Editor",
    "extensions": [
      "feature"
    ]
  },
  {
    "filetype": "Gitignore",
    "handler": "Editor",
    "extensions": [
      ".gitignore"
    ]
  },
  {
    "filetype": "Glsl",
    "handler": "Editor",
    "extensions": [
      "glsl",
      "frag",
      "vert"
    ]
  },
  {
    "filetype": "Gobstones",
    "handler": "Editor",
    "extensions": [
      "gbs"
    ]
  },
  {
    "filetype": "go",
    "handler": "Editor",
    "extensions": [
      "golang"
    ]
  },
  {
    "filetype": "GraphQLSchema",
    "handler": "Editor",
    "extensions": [
      "gql"
    ]
  },
  {
    "filetype": "Groovy",
    "handler": "Editor",
    "extensions": [
      "groovy"
    ]
  },
  {
    "filetype": "HAML",
    "handler": "Editor",
    "extensions": [
      "haml"
    ]
  },
  {
    "filetype": "Handlebars",
    "handler": "Editor",
    "extensions": [
      "hbs",
      "handlebars",
      "mustache"
    ]
  },
  {
    "filetype": "Smarty",
    "handler": "Editor",
    "extensions": [
      "tpl",
      "smarty"
    ]
  },
  {
    "filetype": "Haskell",
    "handler": "Editor",
    "extensions": [
      "hs"
    ]
  },
  {
    "filetype": "Haskell_Cabal",
    "handler": "Editor",
    "extensions": [
      "cabal"
    ]
  },
  {
    "filetype": "haXe",
    "handler": "Editor",
    "extensions": [
      "hx"
    ]
  },
  {
    "filetype": "Hjson",
    "handler": "Editor",
    "extensions": [
      "hjson"
    ]
  },
  {
    "filetype": "HTML_Elixir",
    "handler": "Editor",
    "extensions": [
      "eex",
      "html.eex"
    ]
  },
  {
    "filetype": "HTML_Ruby",
    "handler": "Editor",
    "extensions": [
      "erb",
      "rhtml",
      "html.erb"
    ]
  },
  {
    "filetype": "INI",
    "handler": "Editor",
    "extensions": [
      "ini",
      "conf",
      "cfg",
      "prefs"
    ]
  },
  {
    "filetype": "Io",
    "handler": "Editor",
    "extensions": [
      "io"
    ]
  },
  {
    "filetype": "Jack",
    "handler": "Editor",
    "extensions": [
      "jack"
    ]
  },
  {
    "filetype": "Jade",
    "handler": "Editor",
    "extensions": [
      "jade",
      "pug"
    ]
  },
  {
    "filetype": "Java",
    "handler": "Editor",
    "extensions": [
      "java"
    ]
  },
  {
    "filetype": "JavaScript",
    "handler": "Editor",
    "extensions": [
      "js",
      "jsm"
    ]
  },
  {
    "filetype": "JSX",
    "handler": "Editor",
    "extensions": [
      "jsx"
    ]
  },
  {
    "filetype": "JSON",
    "handler": "Editor",
    "extensions": [
      "json"
    ]
  },
  {
    "filetype": "JSONiq",
    "handler": "Editor",
    "extensions": [
      "jq"
    ]
  },
  {
    "filetype": "JSP",
    "handler": "Editor",
    "extensions": [
      "jsp"
    ]
  },
  {
    "filetype": "JSSM",
    "handler": "Editor",
    "extensions": [
      "jssm",
      "jssm_state"
    ]
  },
  {
    "filetype": "Julia",
    "handler": "Editor",
    "extensions": [
      "jl"
    ]
  },
  {
    "filetype": "Kotlin",
    "handler": "Editor",
    "extensions": [
      "kt",
      "kts"
    ]
  },
  {
    "filetype": "Tex",
    "handler": "Editor",
    "extensions": [
      "tex"
    ]
  },
  {
    "filetype": "LaTeX",
    "handler": "Editor",
    "extensions": [
      "latex",
      "ltx",
      "bib"
    ]
  },
  {
    "filetype": "LESS",
    "handler": "Editor",
    "extensions": [
      "less"
    ]
  },
  {
    "filetype": "Liquid",
    "handler": "Editor",
    "extensions": [
      "liquid"
    ]
  },
  {
    "filetype": "Lisp",
    "handler": "Editor",
    "extensions": [
      "lisp"
    ]
  },
  {
    "filetype": "LiveScript",
    "handler": "Editor",
    "extensions": [
      "ls"
    ]
  },
  {
    "filetype": "LogiQL",
    "handler": "Editor",
    "extensions": [
      "logic",
      "lql"
    ]
  },
  {
    "filetype": "LSL",
    "handler": "Editor",
    "extensions": [
      "lsl"
    ]
  },
  {
    "filetype": "Lua",
    "handler": "Editor",
    "extensions": [
      "lua"
    ]
  },
  {
    "filetype": "LuaPage",
    "handler": "Editor",
    "extensions": [
      "lp"
    ]
  },
  {
    "filetype": "Lucene",
    "handler": "Editor",
    "extensions": [
      "lucene"
    ]
  },
  {
    "filetype": "Makefile",
    "handler": "Editor",
    "extensions": [
      ".Makefile",
      ".GNUmakefile",
      ".makefile",
      ".OCamlMakefile",
      ".make"
    ]
  },
  {
    "filetype": "Markdown",
    "handler": "Editor",
    "extensions": [
      "md",
      "markdown"
    ]
  },
  {
    "filetype": "Mask",
    "handler": "Editor",
    "extensions": [
      "mask"
    ]
  },
  {
    "filetype": "MATLAB",
    "handler": "Editor",
    "extensions": [
      "matlab"
    ]
  },
  {
    "filetype": "Maze",
    "handler": "Editor",
    "extensions": [
      "mz"
    ]
  },
  {
    "filetype": "MEL",
    "handler": "Editor",
    "extensions": [
      "mel"
    ]
  },
  {
    "filetype": "MUSHCode",
    "handler": "Editor",
    "extensions": [
      "mc",
      "mush"
    ]
  },
  {
    "filetype": "MySQL",
    "handler": "Editor",
    "extensions": [
      "mysql"
    ]
  },
  {
    "filetype": "Nix",
    "handler": "Editor",
    "extensions": [
      "nix"
    ]
  },
  {
    "filetype": "NSIS",
    "handler": "Editor",
    "extensions": [
      "nsi",
      "nsh"
    ]
  },
  {
    "filetype": "ObjectiveC",
    "handler": "Editor",
    "extensions": [
      "m",
      "mm"
    ]
  },
  {
    "filetype": "OCaml",
    "handler": "Editor",
    "extensions": [
      "ml",
      "mli"
    ]
  },
  {
    "filetype": "Pascal",
    "handler": "Editor",
    "extensions": [
      "pas",
      "p"
    ]
  },
  {
    "filetype": "Perl",
    "handler": "Editor",
    "extensions": [
      "pl",
      "pm"
    ]
  },
  {
    "filetype": "pgSQL",
    "handler": "Editor",
    "extensions": [
      "pgsql"
    ]
  },
  {
    "filetype": "PHP",
    "handler": "Editor",
    "extensions": [
      "php",
      "phtml",
      "shtml",
      "php3",
      "php4",
      "php5",
      "phps",
      "phpt",
      "aw",
      "ctp",
      "module"
    ]
  },
  {
    "filetype": "Pig",
    "handler": "Editor",
    "extensions": [
      "pig"
    ]
  },
  {
    "filetype": "Powershell",
    "handler": "Editor",
    "extensions": [
      "ps1"
    ]
  },
  {
    "filetype": "Praat",
    "handler": "Editor",
    "extensions": [
      "praat",
      "praatscript",
      "psc",
      "proc"
    ]
  },
  {
    "filetype": "Prolog",
    "handler": "Editor",
    "extensions": [
      "plg",
      "prolog"
    ]
  },
  {
    "filetype": "Properties",
    "handler": "Editor",
    "extensions": [
      "properties"
    ]
  },
  {
    "filetype": "Protobuf",
    "handler": "Editor",
    "extensions": [
      "proto"
    ]
  },
  {
    "filetype": "Python",
    "handler": "Editor",
    "extensions": [
      "py"
    ]
  },
  {
    "filetype": "R",
    "handler": "Editor",
    "extensions": [
      "r"
    ]
  },
  {
    "filetype": "Razor",
    "handler": "Editor",
    "extensions": [
      "cshtml",
      "asp"
    ]
  },
  {
    "filetype": "Rdoc",
    "handler": "Editor",
    "extensions": [
      "Rd"
    ]
  },
  {
    "filetype": "Red",
    "handler": "Editor",
    "extensions": [
      "red",
      "reds"
    ]
  },
  {
    "filetype": "RHTML",
    "handler": "Editor",
    "extensions": [
      "Rhtml"
    ]
  },
  {
    "filetype": "RST",
    "handler": "Editor",
    "extensions": [
      "rst"
    ]
  },
  {
    "filetype": "Ruby",
    "handler": "Editor",
    "extensions": [
      "rb",
      "ru",
      "gemspec",
      "rake",
      ".Guardfile",
      ".Rakefile",
      ".Gemfile"
    ]
  },
  {
    "filetype": "Rust",
    "handler": "Editor",
    "extensions": [
      "rs"
    ]
  },
  {
    "filetype": "SASS",
    "handler": "Editor",
    "extensions": [
      "sass"
    ]
  },
  {
    "filetype": "SCAD",
    "handler": "Editor",
    "extensions": [
      "scad"
    ]
  },
  {
    "filetype": "Scala",
    "handler": "Editor",
    "extensions": [
      "scala"
    ]
  },
  {
    "filetype": "Scheme",
    "handler": "Editor",
    "extensions": [
      "scm",
      "sm",
      "rkt",
      "oak",
      "scheme"
    ]
  },
  {
    "filetype": "SCSS",
    "handler": "Editor",
    "extensions": [
      "scss"
    ]
  },
  {
    "filetype": "SH",
    "handler": "Editor",
    "extensions": [
      "sh",
      "bash",
      ".bashrc"
    ]
  },
  {
    "filetype": "SJS",
    "handler": "Editor",
    "extensions": [
      "sjs"
    ]
  },
  {
    "filetype": "snippets",
    "handler": "Editor",
    "extensions": [
      "snippets"
    ]
  },
  {
    "filetype": "Soy_Template",
    "handler": "Editor",
    "extensions": [
      "soy"
    ]
  },
  {
    "filetype": "Space",
    "handler": "Editor",
    "extensions": [
      "space"
    ]
  },
  {
    "filetype": "SQL",
    "handler": "Editor",
    "extensions": [
      "sql"
    ]
  },
  {
    "filetype": "SQLServer",
    "handler": "Editor",
    "extensions": [
      "sqlserver"
    ]
  },
  {
    "filetype": "Stylus",
    "handler": "Editor",
    "extensions": [
      "styl",
      "stylus"
    ]
  },
  {
    "filetype": "SVG",
    "handler": "Editor",
    "extensions": [
      "svg"
    ]
  },
  {
    "filetype": "Swift",
    "handler": "Editor",
    "extensions": [
      "swift"
    ]
  },
  {
    "filetype": "Tcl",
    "handler": "Editor",
    "extensions": [
      "tcl"
    ]
  },
  {
    "filetype": "Text",
    "handler": "Editor",
    "extensions": [
      "txt"
    ]
  },
  {
    "filetype": "Textile",
    "handler": "Editor",
    "extensions": [
      "textile"
    ]
  },
  {
    "filetype": "Toml",
    "handler": "Editor",
    "extensions": [
      "toml"
    ]
  },
  {
    "filetype": "TSX",
    "handler": "Editor",
    "extensions": [
      "tsx"
    ]
  },
  {
    "filetype": "Twig",
    "handler": "Editor",
    "extensions": [
      "twig",
      "swig"
    ]
  },
  {
    "filetype": "Typescript",
    "handler": "Editor",
    "extensions": [
      "ts",
      "typescript",
      "str"
    ]
  },
  {
    "filetype": "Vala",
    "handler": "Editor",
    "extensions": [
      "vala"
    ]
  },
  {
    "filetype": "VBScript",
    "handler": "Editor",
    "extensions": [
      "vbs",
      "vb"
    ]
  },
  {
    "filetype": "Velocity",
    "handler": "Editor",
    "extensions": [
      "vm"
    ]
  },
  {
    "filetype": "Verilog",
    "handler": "Editor",
    "extensions": [
      "v",
      "vh",
      "sv",
      "svh"
    ]
  },
  {
    "filetype": "VHDL",
    "handler": "Editor",
    "extensions": [
      "vhd",
      "vhdl"
    ]
  },
  {
    "filetype": "Wollok",
    "handler": "Editor",
    "extensions": [
      "wlk",
      "wpgm",
      "wtest"
    ]
  },
  {
    "filetype": "XML",
    "handler": "Editor",
    "extensions": [
      "xml",
      "rdf",
      "rss",
      "wsdl",
      "xslt",
      "atom",
      "mathml",
      "mml",
      "xul",
      "xbl",
      "xaml"
    ]
  },
  {
    "filetype": "XQuery",
    "handler": "Editor",
    "extensions": [
      "xq"
    ]
  },
  {
    "filetype": "YAML",
    "handler": "Editor",
    "extensions": [
      "yaml",
      "yml"
    ]
  }
]