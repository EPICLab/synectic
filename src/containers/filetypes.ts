import * as io from './io';
import { PathLike } from 'fs-extra';

const FILETYPES_PATH = './src/containers/filetypes.json';

export type Filetype = {
  name: string;
  handler: string;
  extensions: string[];
}

/**
 * Asynchronously locate and extract supported filetype metadata based on file extension.
 * @param extension A file extension for a supported language.
 * * @param filetypesPath Optional path to a JSON file containing filetype metadata.
 * @return A Promise object for a Filetype corresponding to the extension, or undefined if extension is unsupported.
 */
export const findExtensionType = (extension: string, filetypesPath: PathLike = FILETYPES_PATH): Promise<Filetype | undefined> => {
  return new Promise((resolve, reject) => {
    io.readFileAsync(filetypesPath)
      .then(content => io.deserialize<Filetype[]>(content))
      .then(filetypes => resolve(filetypes.find(f => f.extensions.some(e => e === extension))))
      .catch(error => reject(error.message));
  });
}