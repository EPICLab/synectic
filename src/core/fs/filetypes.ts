import * as io from './io';
import { snackbar } from './notifications';

export interface Filetype {
  name: string;
  handler: string;
  extensions: string[];
}

/**
 * Asynchronous search for defined filetype corresponding to a filetype name.
 * @param name The file format name to search for within filetypes.
 * @return Filetype instance corresponding to name, or undefined if no matches found.
 */
export function searchName(name: string): Promise<Filetype | undefined> {
  return new Promise((resolve, reject) => {
    io.readFileAsync('src/core/fs/filetypes.json')
      .then(content => io.deserialize<Filetype[]>(content))
      .then(filetypes => resolve(filetypes.find(f => f.name.indexOf(name) != -1)))
      .catch(error => reject(snackbar(global.Synectic.current, error.message, 'Filetype Search Error: Names')));
  });
}

/**
 * Asynchronous search for defined filetype corresponding to a file extension.
 * @param extension The file extension to search for within filetypes.
 * @return Filetype instance corresponding to extension, or undefined if no matches found.
 */
export function searchExt(extension: string): Promise<Filetype | undefined> {
  return new Promise((resolve, reject) => {
    io.readFileAsync('src/core/fs/filetypes.json')
      .then(content => io.deserialize<Filetype[]>(content))
      .then(filetypes => resolve(filetypes.find(f => f.extensions.some(e => e === extension))))
      .catch(error => reject(snackbar(global.Synectic.current, error.message, 'Filetype Search Error: Extensions')));
    });
}
