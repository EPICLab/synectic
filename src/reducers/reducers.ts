import { combineReducers } from 'redux';
import {
  // ADD_TODO,
  // TOGGLE_TODO,
  SET_VISIBILITY_FILTER,
  VisibilityFilters,
  ActionType
} from '../actions/actions';

const { SHOW_ALL } = VisibilityFilters;

// type TodoItem = {
//   text: string;
//   completed: boolean;
// }

// type TodoState = {
//   visibilityFilter: string;
//   todos: TodoItem[];
// }

function visibilityFilter(state = SHOW_ALL, action: ActionType) {
  switch (action.type) {
    case SET_VISIBILITY_FILTER:
      return action.filter;
    default:
      return state;
  }
}

// function todos(state: TodoItem[] = [], action: ActionType) {
//   switch (action.type) {
//     case ADD_TODO:
//       return [
//         ...state,
//         {
//           text: action.text,
//           completed: false
//         }
//       ];
//     case TOGGLE_TODO:
//       return state.map((todo, index) => {
//         if (index === action.index) {
//           return Object.assign({}, todo, {
//             completed: !todo.completed
//           });
//         }
//         return todo;
//       });
//     default:
//       return state;
//   }
// }

const todoApp = combineReducers({
  visibilityFilter
});

export default todoApp;