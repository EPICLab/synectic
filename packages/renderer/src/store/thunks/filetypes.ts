import {extractExtension, extractStats, uuid} from '#preload';
import type {PathLike} from '@syn-types/app';
import type {Filetype} from '@syn-types/filetype';
import {createAppAsyncThunk} from '../hooks';
import filetypeSelectors from '../selectors/filetypes';
import {filetypeAdded} from '../slices/filetypes';

export const fetchFiletype = createAppAsyncThunk<Filetype | undefined, PathLike>(
  'filetypes/fetchFiletype',
  async (filepath, thunkAPI) => {
    const state = thunkAPI.getState();
    const stats = await extractStats(filepath);
    const result = stats?.isDir
      ? filetypeSelectors.selectByFiletype(state, 'Directory')
      : filetypeSelectors.selectByExtension(state, extractExtension(filepath));
    return result.length > 0 ? result[0] : filetypeSelectors.selectByFiletype(state, 'Text')[0];
  },
);

export const importFiletypes = createAppAsyncThunk<void, Omit<Filetype, 'id'>[]>(
  'filetypes/importFiletypes',
  async (filetypes, thunkAPI) => {
    filetypes.map(filetype => {
      thunkAPI.dispatch(filetypeAdded({id: uuid(), ...filetype}));
    });
  },
);
