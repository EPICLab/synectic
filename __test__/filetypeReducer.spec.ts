import type { Filetype } from '../src/types';
import { filetypeReducer } from '../src/store/reducers/filetypes';
import { ActionKeys } from '../src/store/actions';

describe('filetypeReducer', () => {
  const filetypes: { [id: string]: Filetype } = {
    '67': {
      id: '67',
      filetype: 'Python',
      handler: 'Editor',
      extensions: ['py']
    }
  }

  const newFiletype: Filetype = {
    id: '55',
    filetype: 'JavaScript',
    handler: 'Editor',
    extensions: ['js', 'jsm']
  }

  it('filetypeReducer returns default state when current state is blank', () => {
    const newFiletypes = filetypeReducer(undefined, { type: ActionKeys.ADD_FILETYPE, id: newFiletype.id, filetype: newFiletype });
    expect(Object.keys(newFiletypes)).toHaveLength(1);
    expect(newFiletypes).toMatchSnapshot();
  });

  it('filetypeReducer appends a new filetype to state on action ADD_FILETYPE', () => {
    const newFiletypes = filetypeReducer(filetypes, { type: ActionKeys.ADD_FILETYPE, id: newFiletype.id, filetype: newFiletype });
    expect(Object.keys(newFiletypes)).toHaveLength(2);
    expect(newFiletypes).toMatchSnapshot();
  });

  it('filetypeReducer removes a filetype from state on action REMOVE_FILETYPE', () => {
    const matchedFiletypes = filetypeReducer(filetypes, { type: ActionKeys.REMOVE_FILETYPE, id: '67' });
    expect(Object.keys(matchedFiletypes)).toHaveLength(0);
  });

  it('filetypeReducer resolves non-matching filetype in state on action REMOVE_FILETYPE', () => {
    const nonMatchedFiletypes = filetypeReducer(filetypes, { type: ActionKeys.REMOVE_FILETYPE, id: newFiletype.id });
    expect(Object.keys(nonMatchedFiletypes)).toHaveLength(Object.keys(filetypes).length);
  });

  it('filetypeReducer updates state of matched filetype on action UPDATE_FILETYPE', () => {
    const targetFiletype = filetypes['67'];
    const updatedFiletypes = filetypeReducer(filetypes, {
      type: ActionKeys.UPDATE_FILETYPE, id: targetFiletype.id, filetype: {
        extensions: [...targetFiletype.extensions, 'pym']
      }
    });
    expect(updatedFiletypes).not.toMatchObject(filetypes);
    expect(updatedFiletypes).toMatchSnapshot();
  });
})