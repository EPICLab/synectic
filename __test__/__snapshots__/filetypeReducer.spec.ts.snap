// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filetypeReducer filetypeReducer appends a new filetype to state on action ADD_FILETYPE 1`] = `
Object {
  "55": Object {
    "extensions": Array [
      "js",
      "jsm",
    ],
    "filetype": "JavaScript",
    "handler": "Editor",
    "id": "55",
  },
  "67": Object {
    "extensions": Array [
      "py",
    ],
    "filetype": "Python",
    "handler": "Editor",
    "id": "67",
  },
}
`;

exports[`filetypeReducer filetypeReducer returns default state when current state is blank 1`] = `
Object {
  "55": Object {
    "extensions": Array [
      "js",
      "jsm",
    ],
    "filetype": "JavaScript",
    "handler": "Editor",
    "id": "55",
  },
}
`;

exports[`filetypeReducer filetypeReducer updates state of matched filetype on action UPDATE_FILETYPE 1`] = `
Object {
  "67": Object {
    "extensions": Array [
      "py",
      "pym",
    ],
    "filetype": "Python",
    "handler": "Editor",
    "id": "67",
  },
}
`;
